src/main/java/com/suse/salt/netapi/datatypes/target/DictionaryTarget.java

```diff
@@ -0,0 +1,45 @@
+package com.suse.salt.netapi.datatypes.target;
+
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Base class for all grain and pillar target types
+ */
+abstract class DictionaryTarget extends AbstractTarget<String> {
+
+    protected final DictionaryTargetExpression expr;
+
+    public DictionaryTarget(DictionaryTargetExpression expr) {
```

Comment:
That is what I had in 40efc77c4. I moved to DictionaryTargetExpression because I found I was duplicating a lot of code. You have 5 types that all take the same expression format, and each of them has 4 constructors. So I had 20 constructors that just called the base class constructor with the same arguments, and duplicated tests for all of these as well.
Does that sound like the right way to solve that?

Reply:
I don't see a better way to solve that but at least those constructors won't have any logic in them so not much to get wrong. Its annoying to have this kind of boilerplate but i think its worth it if it improves usability of the API.